tsdb/wlog/watcher.go

```diff
@@ -296,9 +294,17 @@ func (w *Watcher) Run() error {
 		w.currentSegmentMetric.Set(float64(currentSegment))
 		level.Debug(w.logger).Log("msg", "Processing segment", "currentSegment", currentSegment)
 
+		// Reset the value of lastSegment each iteration, this is to avoid having to wait too long for
+		// between reads if we're reading a segment that is not the most recent segment after startup.
+		_, lastSegment, err = w.firstAndLast()
+		if err != nil {
+			return fmt.Errorf("wal.Segments: %w", err)
+		}
+		tail := currentSegment >= lastSegment
```

Comment:
I don't know if this was intentional (I cannot see it explicitly stated)

But I think that in order to help the watcher catch up, we started making it skip samples records now (`tail=false` here passed to `readSegment` makes it skip/ignore the samples records)

When this issue was discussed, the `case <-segmentTicker.C` block https://github.com/prometheus/prometheus/blob/8849b7dcadc2960c70467734efce35d737be25f5/tsdb/wlog/watcher.go#L451-L461 was ignored, even though it seems to be there for this same reason: "read the current segment until the end if a new one is there" and it checks that every `segmentCheckPeriod = 100 * time.Millisecond` which should be sufficient.

Reply:
I added a check that shows the samples being dropped here https://github.com/prometheus/prometheus/pull/14434

Reply:
I propose that the reason for the bug is that `tail` means several things in the code.
I will try a PR where the meanings are broken out into separate parameters.

Reply:
Yep,I think a `seriesRecordsOnly` as a param to `readSegment` makes more sense, or make `readSegment` reads everything and have a `readSegmentToBuildSeries` or sth to only build series, is clearer.

Yep, I have a local branch with some cleaning regarding that, I was planning to push with the fix into https://github.com/prometheus/prometheus/pull/14434, but, please, be my guest.

Regarding the bug fix, I suggest we only revert https://github.com/prometheus/prometheus/pull/13583 and re-open https://github.com/prometheus/prometheus/issues/13471 to learn more about the real issue.

Strangely `TestRun_AvoidNotifyWhenBehind` still passes after the revert, locally.

I think running that operation every `segmentCheckPeriod = 100 * time.Millisecond` is acceptable.

Reply:
I'm not sure I agree that just the new segment check via the ticker, without some additional changes in other places, is enough. This [line](https://github.com/prometheus/prometheus/blob/8849b7dcadc2960c70467734efce35d737be25f5/tsdb/wlog/watcher.go#L314) would have had to change to be something like:

```
_, lastSegment, err := w.firstAndLast()
...
currentSegment = lastSegment
```

But then we introduce yet another way to lose samples.

I agree with @bboreham that the overloading of the meaning of `tail` is the problem. We still want to check for segments on some interval, read from the currently written to segment _only_ when we're notified that it's been written to (or on some timer just in case write events are lost within RW somewhere), be notified of checkpoints, etc.

I think we could skip the select statement within `watch` completely if `watch` is not currently reading the segment that is being written to, we really want to just read and send the samples from `currentSegment` as fast as possible in that case (based on the available resources and queue config).

Reply:
I don’t have any stats/data to back my statements, I was just suggesting that unless we have proof (I don't think https://github.com/prometheus/prometheus/issues/13471 provided any) that setting `segmentCheckPeriod = 100 * time.Millisecond` could be problematic (reduces read throughput) in certain setups, reverting to the old behavior is the safest/simplest solution.

Reply:
Hi, sorry I didn't follow this conversation and pushed my own ideas at #14439 

I think it would be ok as a backport for v2.52 to revert and fix just the bit that is wrong.